### PR TITLE
repo dir trash API

### PR DESCRIPTION
**Pull request by sniper-py**

### CODE REVIEW

File: seahub/urls.py [hunk 1]
```diff
@@ -326,6 +326,7 @@
     url(r'^api/v2.1/repos/(?P<repo_id>[-0-9a-f]{36})/dir/$', DirView.as_view(), name='api-v2.1-dir-view'),
     url(r'^api/v2.1/repos/(?P<repo_id>[-0-9a-f]{36})/dir/detail/$', DirDetailView.as_view(), name='api-v2.1-dir-detail-view'),
     url(r'^api/v2.1/repos/(?P<repo_id>[-0-9a-f]{36})/trash/$', RepoTrash.as_view(), name='api-v2.1-repo-trash'),
+    url(r'^api/v2.1/repos/(?P<repo_id>[-0-9a-f]{36})/trash/dir/$', RepoDirTrash.as_view(), name='api-v2.1-repo-dir-trash'),
```
[the remark this answers is not in the document]
Comment (imwhatiam) [reply seahub/urls.py]: url(r'^api/v2.1/repos/(?P<repo_id>[-0-9a-f]{36})/commits/{commit_id}dir/$', RepoCommitDirView.as_view(), name='api-v2.1-dir-view'),

File: seahub/api2/endpoints/repo_trash.py [hunk 2]
```diff
@@ -164,3 +164,72 @@ def delete(self, request, repo_id, format=None):
...
+        dir_entries = seafile_api.list_dir_by_commit_and_path(repo_id, commit_id, path)
+        if dir_entries is None:
+            error_msg = 'Folder %s not found.' % path
+            return api_error(status.HTTP_404_NOT_FOUND, error_msg)
+
+        items = []
+        for dirent in dir_entries:
+            items.append(self._get_item_info(dirent, path))
+
+        return Response({'data': items})
```
[the remark this answers is not in the document]
Comment (imwhatiam) [reply seahub/api2/endpoints/repo_trash.py]: https://dev.seafile.com/seahub/lib/663d57b8-1602-4d6c-a8e3-1b237ca3a766/file/%E8%A7%84%E8%8C%83/seahub%20web%20api%20%E8%A7%84%E8%8C%83.md

File: seahub/api2/endpoints/repo_trash.py [hunk 3]
```diff
@@ -164,3 +164,72 @@ def delete(self, request, repo_id, format=None):
...
+
+    authentication_classes = (TokenAuthentication, SessionAuthentication)
+    permission_classes = (IsAuthenticated, )
+    throttle_classes = (UserRateThrottle, )
+
+    def _get_item_info(self, item, path):
+
+        item_info = {
+            'parent_dir': path,
+            'obj_name': item.obj_name,
```
Comment (imwhatiam) [inline seahub/api2/endpoints/repo_trash.py]: obj_name -> name

File: seahub/api2/endpoints/repo_trash.py [hunk 4]
```diff
@@ -164,3 +164,72 @@ def delete(self, request, repo_id, format=None):
...
+        }
+
+        if stat.S_ISDIR(item.mode):
+            is_dir = True
+        else:
+            is_dir = False
+
+        item_info['is_dir'] = is_dir
```
Comment (imwhatiam) [inline seahub/api2/endpoints/repo_trash.py]: Request URL: https://dev.seafile.com/seahub/api/v2.1/repos/201395ac-c152-4eca-84e7-dece32b4e654/dir/?p=%2F&with_thumbnail=true

File: seahub/api2/endpoints/repo_trash.py [hunk 5]
```diff
@@ -164,3 +164,72 @@ def delete(self, request, repo_id, format=None):
...
+            error_msg = 'Library %s not found.' % repo_id
+            return api_error(status.HTTP_404_NOT_FOUND, error_msg)
+
+        commit = seafile_api.get_commit(repo.id, repo.version, commit_id)
+        if not commit:
+            error_msg = 'Commit %s not found.' % commit
+            return api_error(status.HTTP_404_NOT_FOUND, error_msg)
+
+        # permission check
+        if check_folder_permission(request, repo_id, '/') is None:
```
Comment (imwhatiam) [inline seahub/api2/endpoints/repo_trash.py]: not ....

Comment (imwhatiam) [reply seahub/api2/endpoints/repo_trash.py]: TODO,   r or rw

File: seahub/api2/endpoints/repo_trash.py [hunk 6]
```diff
@@ -164,3 +164,72 @@ def delete(self, request, repo_id, format=None):
...
+            error_msg = 'Permission denied.'
+            return api_error(status.HTTP_403_FORBIDDEN, error_msg)
+
+        dir_entries = seafile_api.list_dir_by_commit_and_path(repo_id, commit_id, path)
+        if dir_entries is None:
```
Comment (imwhatiam) [inline seahub/api2/endpoints/repo_trash.py]: 1, get dir id
2, list dir by dir id

File: seahub/api2/endpoints/repo_commit_dir.py [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+        if not dir_id:
+            error_msg = 'Folder %s not found.' % path
+            return api_error(status.HTTP_404_NOT_FOUND, error_msg)
+
+        # permission check
+        if check_folder_permission(request, repo_id, '/') != 'rw':
+            error_msg = 'Permission denied.'
+            return api_error(status.HTTP_403_FORBIDDEN, error_msg)
+
+        dir_entries = seafile_api.list_dir_by_dir_id(repo_id, dir_id)
```
Comment (imwhatiam) [inline seahub/api2/endpoints/repo_commit_dir.py]: list dir by commit and path

File: seahub/api2/endpoints/repo_commit_dir.py [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+        if stat.S_ISDIR(item.mode):
+            item_info['type'] = 'dir'
+        else:
+            item_info['type'] = 'file'
+            item_info['size'] = item.size
+
+        return item_info
+
+    def get(self, request, repo_id, commit_id, format=None):
+        """ Return files/dirs in a trash or history dir
```
Comment (imwhatiam) [inline seahub/api2/endpoints/repo_commit_dir.py]: list dir by commit
used when get files/dirs in a trash or history dir

File: seahub/api2/endpoints/repo_commit_dir.py [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+            error_msg = 'Commit %s not found.' % commit
+            return api_error(status.HTTP_404_NOT_FOUND, error_msg)
+
+        dir_id = seafile_api.get_dir_id_by_commit_and_path(repo_id, commit_id, path)
+        if not dir_id:
+            error_msg = 'Folder %s not found.' % path
+            return api_error(status.HTTP_404_NOT_FOUND, error_msg)
+
+        # permission check
+        if check_folder_permission(request, repo_id, '/') != 'rw':
```
Comment (imwhatiam) [inline seahub/api2/endpoints/repo_commit_dir.py]: TODO

File: tests/api/endpoints/test_repo_commit_dir.py [hunk 10]
```diff
@@ -0,0 +1,68 @@
...
+        self.repo_id = self.repo.id
+        self.repo_name = self.repo.repo_name
+
+        self.file_path = self.file
+        self.file_name = os.path.basename(self.file_path)
+
+        self.folder_path = self.folder
+        self.folder_name = os.path.basename(self.folder.rstrip('/'))
+
+        self.commit_id_url = reverse('api-v2.1-repo-trash', args=[self.repo_id])
```
Comment (imwhatiam) [inline tests/api/endpoints/test_repo_commit_dir.py]: trash url

File: seahub/api2/endpoints/repo_commit_dir.py (hunk 9 above)
Comment (llj) [inline seahub/api2/endpoints/repo_commit_dir.py]: path 不会一直是 '/'

File: seahub/api2/endpoints/repo_commit_dir.py [hunk 11]
```diff
@@ -0,0 +1,82 @@
...
+
+    def get(self, request, repo_id, commit_id, format=None):
+        """ Return files/dirs in a trash or history dir
+
+        Permission checking:
+        1. all authenticated user can perform this action.
+        """
+
+        # argument check
+        path = request.GET.get('path', None)
```
Comment (imwhatiam) [inline seahub/api2/endpoints/repo_commit_dir.py]: '/'

**Pull request merged**